File: oop/python/derek/ride_share/ride_share_company.py

from typing import Optional, Dict, List
from driver import Driver
from passenger import Passenger

class RideShareCompany:

  def __init__(self, name) -> None:
    self.name = name
    self.drivers: Dict[str, Driver] = {}
    self.passengers: Dict[str, Passenger] = {}
    self.free_drivers: List[str] = []
    self.current_rides: List[str, str] = {}

  def register_driver(self, name: str) -> Optional[str]:
    if (name in self.drivers):
      return "Driver already registered."
    self.drivers[name] = Driver(name)
    self.free_drivers.append(name)

  def register_passenger(self, name: str) -> Optional[str]:
    if (name in self.passengers):
      return "Passenger already registered."
    self.passengers[name] = Passenger(name)

  def request_ride(self, passenger_name: str) -> Optional[str]:
    if (passenger_name not in self.passengers):
      return "Passenger not registered."
    if (len(self.free_drivers) == 0):
      return "No drivers available."
    if (passenger_name in self.current_rides):
      return "Ride already requested."
    
    driver_name = self.free_drivers.pop(0)
    self.drivers[driver_name].start_ride(passenger_name)
    self.passengers[passenger_name].start_ride(driver_name)
    self.current_rides[passenger_name] = driver_name

  def complete_ride(self, passenger_name):
    if (passenger_name not in self.passengers):
      return "Passenger not registered."
    if (passenger_name not in self.current_rides):
      return "No ongoing ride."
    
    driver_name = self.current_rides[passenger_name]
    self.passengers[passenger_name].ride_completed()
    self.drivers[driver_name].ride_completed()
    self.free_drivers.append(driver_name)
    del self.current_rides[passenger_name]

  def get_top_drivers(self, num_drivers: int) -> List[Driver]:
    drivers = self.drivers.values()
    drivers = sorted(drivers, key= lambda driver: (-driver.total_ride, driver.name))
    return drivers[0 : num_drivers]

  def get_top_passengers(self, num_passengers: int) -> List[Passenger]:
    passengers = self.passengers.values()
    passengers = sorted(passengers, key= lambda passenger: (-passenger.total_ride, passenger.name))
    return passengers[0 : num_passengers]

File: oop/typescript/rideShare/RideShareCompany.ts
import { Driver } from "./Driver";
import { Passenger } from "./Passenger";


export class RideShareCompany {
  private name: string;
  private drivers: Map<string, Driver>;
  private passengers: Map<string, Passenger>;
  private freeDrivers: string[];
  private currentRides: Map<string, string>;

  constructor(name: string) {
    this.name = name;
    this.drivers = new Map<string, Driver>();
    this.passengers = new Map<string, Passenger>();
    this.freeDrivers = [];
    this.currentRides = new Map<string, string>();
  }

  registerDriver = (name: string): string | null => {
    if (this.drivers.has(name)) {
      return "Driver already registered";
    }
    this.drivers.set(name, new Driver(name));
    this.freeDrivers.push(name);

    return null;
  }

  registerPassenger = (name: string): string | null => {
    if (this.passengers.has(name)) {
      return "Passenger already registered";
    }
    this.passengers.set(name, new Passenger(name));

    return null;
  }

  requestRide = (passengerName: string): string | null => {
    const passenger: Passenger | undefined = this.passengers.get(passengerName);
    if (!passenger) {
      return "Passenger NOT registered";
    }
    if (this.freeDrivers.length === 0) {
      return "No driver available";
    }
    if (this.currentRides.has(passengerName)) {
      return "Ride already requested";
    }

    const driverName: string = this.freeDrivers.shift() ?? "";
    this.currentRides.set(passengerName, driverName);
    passenger.startRide(driverName);
    this.drivers.get(driverName)?.startRide(passengerName);

    return null;
  }

  completeRide = (passengerName: string): string | null => {
    const passenger: Passenger | undefined = this.passengers.get(passengerName);
    if (!passenger) {
      return "Passenger not registered";
    }
    const driverName: string | undefined = this.currentRides.get(passengerName);
    if (!driverName) {
      return "No ongoing ride";
    }

    this.currentRides.delete(passengerName);
    this.freeDrivers.push(driverName);
    passenger.completeRide();
    this.drivers.get(driverName)?.completeRide();

    return null;
  }

  getTopDrivers = (num: number) => {
    const drivers: Driver[] = Array.from(this.drivers.values());
    drivers.sort((a: Driver, b: Driver) => {
      if (a.getTotalRides() === b.getTotalRides()) {
        return a.getName().localeCompare(b.getName());
      }
      return b.getTotalRides() - a.getTotalRides();
    })
    return drivers.slice(0, Math.min(num, drivers.length));
  }

  getTopPassengers = (num: number) => {
    const passengers: Passenger[] = Array.from(this.passengers.values());
    passengers.sort((a: Passenger, b: Passenger) => {
      if (a.getTotalRides() === b.getTotalRides()) {
        return a.getName().localeCompare(b.getName());
      }
      return b.getTotalRides() - a.getTotalRides();
    })
    return passengers.slice(0, Math.min(num, passengers.length));
  }
}